pkg/tests/validator_test.go

```diff
@@ -0,0 +1,381 @@
+
+		slotStartTime := ts.API.TimeProvider().SlotStartTime(ts.CurrentSlot()).UTC()
+		for subslotIdx := uint8(0); subslotIdx < test.validationBlocksPerSlot; subslotIdx++ {
+			for _, node := range ts.Validators() {
+				blockName := fmt.Sprintf("block-%s-%d/%d", node.Name, ts.CurrentSlot(), subslotIdx)
+				latestCommitment := node.Protocol.Engines.Main.Get().Storage.Settings().LatestCommitment().Commitment()
+
+				issuingTime := slotStartTime.
+					Add(subslotDuration * time.Duration(subslotIdx)).
+					Add(time.Duration(subSlotBlockCounter))
+
+				validationBlock := ts.IssueValidationBlockWithHeaderOptions(blockName, node,
+					mock.WithSlotCommitment(latestCommitment),
+					mock.WithIssuingTime(issuingTime),
+					mock.WithStrongParents(tip),
+				)
+
+				tip = validationBlock.ID()
+				subSlotBlockCounter += 1
+			}
+		}
+	}
+
+	ts.Wait(ts.Nodes()...)
+
+	// Determine total stakes required for rewards calculation.
+	var totalStake iotago.BaseToken = 0
+	var totalValidatorStake iotago.BaseToken = 0
+	lo.ForEach(ts.Validators(), func(n *mock.Node) {
+		latestCommittedSlot := ts.DefaultWallet().Node.Protocol.Engines.Main.Get().Storage.Settings().LatestCommitment().Slot()
+		accountData, exists, err := ts.DefaultWallet().Node.Protocol.Engines.Main.Get().Ledger.Account(n.Validator.AccountID, latestCommittedSlot)
+		if err != nil || !exists {
+			t.Fatal(exists, err)
+		}
+
+		totalStake += accountData.ValidatorStake + accountData.DelegationStake
+		totalValidatorStake += accountData.ValidatorStake
+	})
+
+	// Determine the rewards the validators actually got.
+	actualRewards := make(map[iotago.AccountID]iotago.Mana, len(ts.Validators()))
+	claimingEpoch := ts.API.TimeProvider().EpochFromSlot(ts.CurrentSlot())
+
+	for _, validatorAccount := range []string{"Genesis:1", "Genesis:2"} {
+		output := ts.DefaultWallet().Output(validatorAccount)
+		accountID := output.Output().(*iotago.AccountOutput).AccountID
+
+		rewardMana, _, _, err := ts.DefaultWallet().Node.Protocol.Engines.Main.Get().SybilProtection.ValidatorReward(
+			accountID,
+			output.Output().FeatureSet().Staking(),
+			claimingEpoch,
+		)
+		if err != nil {
+			panic(err)
+		}
+
+		actualRewards[accountID] = rewardMana
+	}
+
+	for accountID, actualReward := range actualRewards {
+		lastRewardEpoch := iotago.EpochIndex(len(test.epochPerformanceFactors))
+		rewards := make([]reward, 0, lastRewardEpoch)
+		for epoch := iotago.EpochIndex(0); epoch < lastRewardEpoch; epoch++ {
+			epochPerformanceFactor := test.epochPerformanceFactors[epoch]
+			reward := calculateEpochReward(t, ts, accountID, epoch, epochPerformanceFactor, totalStake, totalValidatorStake)
+			rewards = append(rewards, reward)
+		}
+
+		expectedReward := calculateValidatorReward(t, ts, accountID, rewards, startEpoch, claimingEpoch)
+
+		require.Equal(t, expectedReward, actualReward, "expected reward for account %s to be %d, was %d", accountID, expectedReward, actualReward)
+	}
+}
+
+type reward struct {
```

Comment:
parhaps the type should also be called epochReward, and the slice used as input to the calculateValidatorRewards() should be called epochRewards